Add packet tracing utilities

Tener opened the pull request:
Code extracted from `e` PR after relevant comment
- https://github.com/gravitational/teleport.e/pull/5623#discussion_r1878185971

Tener committed:
Add packet tracing utilities

greedy52 commented on lib/srv/db/common/packettrace/tracing.go:
```diff
@@ -0,0 +1,235 @@
+	return nil
+}
+
+func (t *Trace) SaveAsText(file string) error {
```
nit: godoc

Tener replied:
Added.

greedy52 commented on lib/srv/db/common/packettrace/tracing.go:
```diff
@@ -0,0 +1,235 @@
+	return nil
+}
+
+// SaveToPCAP saves the trace to a single merged pcap file.
```
nit: could we document the pre-req?

Tener replied:
Yes, you are right, this has two runtime dependencies.

Tener replied:
Documented.

zmb3 commented on lib/srv/db/common/packettrace/tracing.go:
```diff
@@ -0,0 +1,235 @@
+	clock      clockwork.Clock
+	runCommand func(name string, arg ...string) ([]byte, error)
+
+	mx sync.Mutex
```
```suggestion
	mu sync.Mutex
```

zmb3 commented on lib/srv/db/common/packettrace/tracing.go:
```diff
@@ -0,0 +1,235 @@
+	}
+	defer os.Remove(filenameHex)
+
+	// Invoke text2pcap to convert the hex dump to PCAP format
```
I assume this isn't used for anything where performance matters, right?

Writing in one format and using an external utility to get to the desired format seems rather inefficient, but I don't have any context on what this is for.

Tener replied:
Nope, this is meant for debugging. I'll add comments about this.

Tener replied:
I've added some disclaimers. 

In practice I don't think this would really be that slow, text2pcap is an efficient C program using `libpcap` under the hood, it may well be faster than equivalent Go code even with the intermediate step. Nonetheless, it shouldn't be used in performance critical contexts, and I've added respective comments.

zmb3 commented on lib/srv/db/common/packettrace/tracing.go:
```diff
@@ -0,0 +1,235 @@
+	return nil
+}
+
+func (t *Trace) SaveAsText(file string) error {
```
Should we implement `io.WriterTo` instead for more flexibility and interoperability?

Tener replied:
I can add this other method and have `SaveAsText` call it.

Tener replied:
Added.

zmb3 commented on lib/srv/db/common/packettrace/tracing_test.go:
```diff
@@ -0,0 +1,175 @@
+	})
+
+	tmp := t.TempDir()
+	fileOut := path.Join(tmp, "trace.txt")
```
```suggestion
	fileOut := filepath.Join(tmp, "trace.txt")
```

espadolini reviewed (commented):
Please rename this to `packetcapture` or something along those lines, "tracing" is already used by distributed tracing in the OpenTelemetry sense and by gravitational/trace for stack tracing.

This would benefit from package docs to describe what this is used for; is this just for testing, or is it actually used in normal code?

Tener commented:
> Please rename this to `packetcapture` or something along those lines, "tracing" is already used by distributed tracing in the OpenTelemetry sense and by gravitational/trace for stack tracing.

I think the name is correct, but I agree it can lead to confusion. I'll find a better one.
 
> This would benefit from package docs to describe what this is used for; is this just for testing, or is it actually used in normal code?

Sure, I'll add some docs. Currently there is a single user for this API, check out the linked PR: https://github.com/gravitational/teleport.e/pull/5623 and this commit in particular https://github.com/gravitational/teleport.e/pull/5623/commits/59a5cefe36a736256138ef63d393e30740563a5b. 

It has been suggested, however, that given generic nature of this code it can be shared more widely. I can see us adopting these utilities in a wider context.

Tener committed:
Rename package, add docs, other minor changes.

Tener committed:
Merge branch 'master' into tener/packet-tracing

Tener commented:
> Please rename this to `packetcapture` or something along those lines, "tracing" is already used by distributed tracing in the OpenTelemetry sense and by gravitational/trace for stack tracing.

Renamed to `packetcapture`, everything that was `trace` is now `capture`.
 
> This would benefit from package docs to describe what this is used for; is this just for testing, or is it actually used in normal code?

Added comments.

espadolini commented on lib/srv/db/common/packetcapture/capture.go:
```diff
@@ -0,0 +1,248 @@
+type Capture struct {
+	packets    []PacketEntry
+	clock      clockwork.Clock
+	runCommand func(name string, arg ...string) ([]byte, error)
```
Add a godoc hint that this is expected to behave like `(*exec.Cmd).CombinedOutput`?

Tener commented on lib/srv/db/common/packetcapture/capture.go:
```diff
@@ -0,0 +1,248 @@
+	return nil
+}
+
+func (c *Capture) WriteTo(w io.Writer) (int64, error) {
```
Hmm, I've added this per Zac's comment above: https://github.com/gravitational/teleport/pull/50053#discussion_r1880507280

As far as I can see there is no contract violation here, the `WriterTo` is documented as follows:

```
// WriterTo is the interface that wraps the WriteTo method.
//
// WriteTo writes data to w until there's no more data to write or
// when an error occurs. The return value n is the number of bytes
// written. Any error encountered during the write is also returned.
//
// The Copy function uses WriterTo if available.
type WriterTo interface {
	WriteTo(w Writer) (n int64, err error)
}
```

Is there something I'm missing here?

Tener replied:
Ugh, not a great fan of tribal-knowledge contracts.

I'll take this escape hatch:
> Of course there may be cases where that doesn't make sense

Coupled by the fact that there is no _requirement_ to modify internal state as a result of `WriteTo()` call.

Tener commented on lib/srv/db/common/packetcapture/capture.go:
```diff
@@ -0,0 +1,248 @@
+	defer os.Remove(filename002)
+
+	// merge two files
+	out, err := c.runCommand(mergecapBin, "-w", baseFilename, filename001, filename002)
```
I don't know what `mergecap` does internally, so for the benefit of simplicity I'd rather avoid that.

The `.pcap` files are rather compact, but there is no builtin constraint. However, we keep all of that data in memory prior to saving, which is likely the more important kind of limitation anyway.

Tener committed:
Update lib/srv/db/common/packetcapture/capture.go

Co-authored-by: Edoardo Spadolini <edoardo.spadolini@goteleport.com>

Tener committed:
add runCommand comment

Tener committed:
Update comments
